README
     Copyright (c) 2010-2016 The SquirrelMail Project Team
     Copyright (c) 2002-2010 Dave McMurtrie


This directory contains the source distribution of SquirrelMail IMAP Proxy.

If you find any bugs in squirrelmail-imap_proxy, please subscribe to the
squirrelmail-imapproxy mailing list and post your bug there.

If you want to subscribe to the squirrelmail-imapproxy mailing list, visit:

https://lists.sourceforge.net/lists/listinfo/squirrelmail-imapproxy

For some Debian specific issues that have come up, read the file README.debian.

To read a little bit about SSL support in squirrelmail-imap_proxy, read
README.ssl.


##############################################################################
HOW TO BUILD
##############################################################################

Before you build squirrelmail-imap_proxy, you're going to need to obtain
the OpenSSL libraries.  Build and install them, then you're ready to move on.

It should be pretty painless:

./configure
make

If you encounter any problems with this, check out README.known_issues to
see if your problem is addressed in there.  If not, ask on the list or file
a bug report.

"make install" will attempt to install the following files:

/usr/local/sbin/in.imapproxyd
/usr/local/sbin/pimpstat

"make install-conf" will attempt to install the sample configuration file:

/etc/imapproxy.conf

"make install-init-systemd" will attempt to install the systemd startup script:

/lib/systemd/system/imap_proxy.service

"make install-init" will attempt to install the startup script and symlinks:

/etc/init.d/imapproxy
/etc/rc2.d/S99imapproxy
/etc/rc0.d/K10imapproxy

"make pkg" will attempt to install a sun package into /var/spool/package.

After you successfully build the proxy server, you'll need to edit the
global configuration file to suit your needs.  By default, the proxy
server looks at /etc/imapproxy.conf as its default configuration file.
You can change this at runtime by supplying the -f argument to
/usr/local/sbin/in.imapproxyd.

/usr/local/sbin/pimpstat is the Polling Imap Mail Proxy Statistical tool.  It's
a curses-based application that you can use to monitor the proxy server.  It's
aware of the same global configuration file as the proxy server, and can also
be told to look at a different configuration file by supplying it as a 
parameter with the -f argument.  Both pimpstat and in.imapproxyd must be using
the same configuration file, since the stat file that they both rely on is
set in the configuration file.


##############################################################################
CONFIGURATION OPTIONS
##############################################################################

server_hostname
---------------
This determines which IMAP server you want to proxy connections to.

cache_size
----------
A slightly misleading configuration option.  It's really the total number of
in-core connections that the server can handle, so it will include all of the
active connections and all of the cached connections.  For performance, no
memory is dynamically allocated within the proxy server once it starts to 
accept connections.  That means that all of the IMAP connection structures have
to be allocated at server startup.  This configuration setting allows you to
control how many of these structures are allocated.  The higher you make this
number, the more memory the server will allocate.  Also, since the proxy
server is multi-threaded it will need to increase the total number of allowed
file descriptors it can have open.  We run with a default setting of 3072
here.  I don't know how big you can make this number before you'd have
problems with setrlimit().

listen_port
-----------
The port that the server binds to and accepts connections on.  This is the
tcp port that IMAP clients will connect to.

cache_expiration_time
---------------------
This is the number of seconds that we keep a connection open to the IMAP server
after a client logs out.  Once a client logs out, the in-core IMAP connection
structure is marked as "cached" instead of "active" and a timestamp is set on
the structure.  Once "cache_expiration_time" seconds have elapsed, this
connection to the server will be closed and any future logins from the user
that was connected on this socket will require a new connection to the server.

proc_username
-------------
The proxy server will NOT run as root.  This is quite simply for security
reasons.  You can specify which user you want the proxy to run as here.

proc_groupname
--------------
This is the groupname that the proxy server will run as.

stat_filename
-------------
The proxy server opens a file and mmap()s a statistical structure when it
starts.  It keeps some really simple numbers in here.  You can use the
pimpstat application to display these numbers and monitor the usage of the
proxy server.

protocol_log_filename
---------------------
The proxy server allows you to turn on protocol logging on a per-user basis.
All proxied traffic for one user will be logged to this file, differentiated
by client or server.  This file is opened at server startup, and is held open
until the server is shut down.

syslog_facility
---------------
The logging facility to be used for all syslog calls.  Any of the possible
facilities listed in the syslog(3C) manpage may be specified here except
for LOG_KERN.

syslog_prioritymask
-------------------
The syslog LOG_UPT() priority mask.  Read more about what this does in the
syslog(3C) manpage.  Any of the possible priorities specified in the
syslog(3C) manpage my be specified here.

send_tcp_keepalives
-------------------
Allows tcp keepalives to be enabled on all sockets if the tcp implementation
supports it.

tls_ca_file
-----------
File containing one or more Certificate Authority (CA) certificates.
See README.ssl for more information.

tls_ca_path
-----------
Path to directory with certificates of CAs.  This directory must have
filenames with the hashed value of the certificate (see openssl(1)).
See README.ssl for more information.

tls_cert_file
-------------
File containing the certificate presented by squirrelmail-imap_proxy to the
server for client authentication during STARTTLS.  See README.ssl for more
information.

tls_key_file
------------
File containing the private key belonging to the client certificate.  See
README.ssl for more information.

tls_verify_server
-----------------
Indicates that the server's TLS certificate must be verified.  See
README.ssl for more information.

tls_ciphers
-----------
Allows you to manually specify TLS cipher suite as documented in openssl
ciphers(1).  See README.ssl for more information.

tls_no_tlsv1
------------
Prevents the usage of TLS v1 (SSL v2 and v3 are already forbidden, which
is hard coded).  See README.ssl for more information.

tls_no_tlsv1.1
--------------
Prevents the usage of TLS v1.1 (SSL v2 and v3 are already forbidden, which
is hard coded).  See README.ssl for more information.

tls_no_tlsv1.2
--------------
Prevents the usage of TLS v1.2 (SSL v2 and v3 are already forbidden, which
is hard coded).  See README.ssl for more information.

force_tls
---------
Requires STARTTLS support by the server (won't use unsecured connections).
See README.ssl for more information.

enable_select_cache
-------------------
Allows SELECT data caching to be enabled or disabled.

foreground_mode
---------------
When enabled, this will prevent squirrelmail-imap_proxy from detaching from
his parent process and controlling terminal.

chroot_directory
----------------
If this value is set to anything, it implicitly causes squirrelmail-imap_proxy
to run in a chroot jail.

enable_admin_commands
---------------------
The internal admin commands that squirrelmail-imap_proxy provides (p_dumpicc,
p_trace, etc) are a security vulnerability if you do not restrict access to
the listen_port using tcp wrappers, iptables, ipchains, ipfilter, an external
firewall, or some other mechanism.  Because of this, the internal admin
commands are disabled by default.


##############################################################################
NEW STATUS RESPONSE:
##############################################################################

Upon a successful LOGIN or AUTHENTICATION, if squirrelmail-imap_proxy reused
an existing connection an untagged IMAP response with the status response
XPROXYREUSE will be issued.  This status response indicates to a client that
any previous initialization does not need to be performed again.

Examples:
---------
1. Login without an active cached connection.

C: a001 LOGIN user password
S: [... untagged response(s) from remote IMAP server ...]
S: a001 OK User logged in

2. Login with an active cached connection.

C: a001 LOGIN user password
S: [... untagged response(s) from remote IMAP server ...]
S: * OK [XPROXYREUSE] IMAP connection reused by squirrelmail-imap_proxy
S: a001 OK User logged in


##############################################################################
ADDITIONAL COMMANDS:
##############################################################################

There are a few additional commands that have been added to allow you to
administer the proxy server.  They're implemented through the same mechanism
as regular IMAP protocol commands, so they require you to telnet to your
proxy server on whatever port you choose to bind to and type them like a
regular protocol transaction.  At our site, we've severely limited access to
the proxy port such that only our webmail machines and one internal admin
machine can access it.  The proxy server software offers no way to limit the
usage of these commands by username so they're not implemented by default.
If you secure the listen_port, you can then set the enable_admin_commands
option in imapproxy.conf to enable the use of these commands.

These commands are available only until you log into the remote IMAP server.

If these commands are available, the string 'XIMAPPROXY' will be listed in
the CAPABILITY string before login/authentication (see RFC 3501 [6.1.1]).

These commands are prefixed with 'X' to comply with RFC 3501 [6.5.1].


XPROXY_NEWLOG
-------------
Since the protocol log file is held open as long as the server is running,
it's difficult to clear the logfile.  This command was added to take care of
that for you.

Usage:
  <tag> XPROXY_NEWLOG

Example:
C: a001 XPROXY_NEWLOG
S: a001 OK Logfile cleared


XPROXY_DUMPICC
--------------
The Dump ICC command allows you to display the internal data structures of the
proxy server.  It can tell you how many connections you currently have open,
what users they're for, and the status of the connections (active or cached).
Use this sparingly if the proxy server is extremely busy.

Usage:
  <tag> XPROXY_DUMPICC

Untagged response(s) (1 for each cached connection):
  * XPROXY_DUMPICC <#> <user> <status>

  <#> = socket descriptor of IMAP connection to backend server
  <user> = Username of connection
  <status> = Either 'Active' or 'Cached'

Example:
C: a001 XPROXY_DUMPICC
S: * XPROXY_DUMPICC 1 foo Cached
S: * XPROXY_DUMPICC 2 foo2 Active
S: a001 OK Completed


XPROXY_TRACE
------------
This is used to turn on or off protocol logging.  If issued without any
arguments, it will disable protocol logging.  If issued with a username
argument, it will turn on logging for that particular user.  Protocol log
output will show up in the file configured as "protocol_log_filename" in the
configuration file.

Usage:
  <tag> XPROXY_TRACE [user]

  [user] = (OPTIONAL) Username to log

Examples:
C: a001 XPROXY_TRACE
S: a001 OK Tracing disabled

C: a001 XPROXY_TRACE foo
S: a001 OK Tracing enabled for user foo


XPROXY_RESETCOUNTERS
--------------------
This allows you to reset the internal counters that pimpstat reports on
without having to restart the server.  These are the counters that are mmap()ed
to the file configured as "stat_filename" in the configuration file.

Usage:
  <tag> XPROXY_RESETCOUNTERS

Example:
C: a001 XPROXY_RESETCOUNTERS
S: a001 OK Counters reset


XPROXY_VERSION
--------------
This is used to show the version number of the IMAP proxy.

Usage:
  <tag> XPROXY_VERSION

Untagged response(s) (1 for each cached connection):
  * XPROXY_VERSION <version>

  <version> = Current version number of squirrelmail-imap_proxy.

Example:
C: a001 XPROXY_VERSION
S: * XPROXY_VERSION 1.2.6
S: a001 OK Completed


ID
--
This is not a command specific to squirrelmail-imap_proxy, rather it
comes from RFC 2971.  It is listed here because the ID command may
be sent before the client authenticates, which means the proxy server
handles it - sort of.  Any ID commands still need to be sent to the
IMAP server, so the proxy responds with a fake "OK" and queues the ID
commands to be sent once a LOGIN or AUTHENTICATE is received from the
client and a server connection is (re)established.

Usage:
  <tag> ID [(parameter list)]

  [(parameter list)] = (OPTIONAL) Parameters to be sent to IMAP server

  See RFC 2971.

Examples:
C: a001 ID
S: a001 OK Completed

C: a001 ID ("x-originating-ip" "10.10.1.2")
S: a001 OK Completed

Responses above from the proxy server are faked and do not necessarily
indicate that the IMAP server will accept the ID commands.  If an ID
command fails when sent to the IMAP server, the proxy server will log
a problem that includes the string "QUEUED PREAUTH failed".

Any ID commands sent by the client AFTER login/authentication are
always sent directly through to the IMAP server and are never handled
by the proxy server.

NOTE: Currently only the last ID command received before client login
is remembered and relayed to the server.  Please let us know if you
find this insufficient.



Happy proxying,

The SquirrelMail Project Team



ChangeLog
2016-09-12  Paul Lesniewski <paul@squirrelmail.org>
	* Added support for systemd startup
	* Added support for accepting pre-auth ID commands (RFC 2971)
	* Added support for DNS RR to cycle through multiple IMAP
	  server backends (thanks to Wolfgang Breyha)
	* Added ability to restrict DNS lookups to IPv4 or IPv6
	  only (thanks to Wolfgang Breyha)
	* Make EGD support conditional, provide compatibility with
	  LibreSSL, other small fixes provided by the BSD team
	* Fixed anomalous crashes recycling used connections (thanks
	  to Emmanuel Dreyfus)
	* Retain compatibility with older OpenSSL versions (thanks to
	  Wolfgang Breyha)
	* Make use of the no_new_privs flag (Linux only) when becoming
	  non-root (thanks to Shawn Landden)

2014-01-20  Paul Lesniewski <paul@squirrelmail.org>
	* Added support for up to TLS v1.2 (thanks to Emmanuel Dreyfus)
	* Added support for ECDHE ciphers (thanks to Emmanuel Dreyfus)
	* Added ability to manually specify TLS ciphers (thanks to Emmanuel Dreyfus)
	* Added server certificate validation (thanks to Emmanuel Dreyfus)

2012-01-01  Paul Lesniewski <paul@squirrelmail.org>
	* Fixed problem where default TLS CA data would never
	  be loaded (thanks to Orion Poplawski)
	* Fixed bug in SSL context initialization

2011-12-16  Paul Lesniewski <paul@squirrelmail.org>
	* Minor Makefile cleanup complements Brad Smith
	* Fixed header include issue for OpenBSD (thanks to Brad Smith)

2011-11-23  Paul Lesniewski <paul@squirrelmail.org>
	* Added DESTDIR support to install scripts (thanks to Brad Smith)

2011-11-09  Paul Lesniewski <paul@squirrelmail.org>
	* Fixed minor bug where too many spaces were added before
	  XIMAPPROXY in the CAPABILITY string (thanks to Kirill Miazine)

2011-04-17  Paul Lesniewski <paul@squirrelmail.org>
	* Added configurable, arbitrary pre-authentication command
	  that the administrator can use to send non-standard
	  commands to the server before each user authenticates
	  (for an example usage, see:
	  http://en.wikipedia.org/wiki/Yahoo!_Mail#Free_IMAP_and_SMTPs_access )

2011-04-17  Paul Lesniewski <paul@squirrelmail.org>
	* Add restart operation to (linux) init script
	* Add BSD-style init script (thanks to Emmanuel Dreyfus)

2011-04-17  Paul Lesniewski <paul@squirrelmail.org>
	* Fixed server connection synchronization issues in the SELECT
	  cache code (ensure server failures result in server connections
	  being fully shut down and removed from connection cache).

2011-04-17  Paul Lesniewski <paul@squirrelmail.org>
	* When NO or BAD response is returned from the server against
	  a LOGIN or AUTHENTICATE request, we now log the full server
	  response and pass it back to the client (useful if client
	  is watching for RFC 5530 response codes).  See:
	  http://thread.gmane.org/gmane.mail.squirrelmail.imapproxy/4

2011-03-11  Paul Lesniewski <paul@squirrelmail.org>
	* Fixed server connection synchronization issue that occurred when
	  clients closed their connection immediately after logging in and
	  associated security issue of closing such server connections
	  without invalidating the connection in the internal connection
	  cache (thanks to Jose Celestino).

2011-03-07  Paul Lesniewski <paul@squirrelmail.org>
	* Added information about how to use squirrelmail-imap_proxy
	  with servers that are only available via imaps

2011-01-16  Paul Lesniewski <paul@squirrelmail.org>
	* Added the ability to authenticate to the IMAP server
	  using SASL plain authentication with fixed authentication
	  credentials (thanks to Jeffrey Arbuckle and Martin Smith)

2010-10-29  Paul Lesniewski <paul@squirrelmail.org>
	* Fixed compilation warnings when openssl is not available.

2010-10-28  Paul Lesniewski <paul@squirrelmail.org>
	* Increased default buffer sizes.
	* Fixed possible buffer overflow issue and a few other small
	  issues, complements Jose Celestino.  Thanks also to Tomas
	  Lindroos.
	* Fixed LOGIN command so that it handles literal arguments
	  correctly.

2010-08-19  Paul Lesniewski <paul@squirrelmail.org>
	* Added several warnings that try to catch the case where someone
	  tries to set the server_port to 993, which is usually a mistaken
	  use of TLS/SSL (port 993 is usually for imaps, which is not
	  supported; IMAP Proxy supports STARTTLS from a "normal" port,
	  which is usually 143)

2010-07-26  Paul Lesniewski <paul@squirrelmail.org>
	* Updated copyrights and other administrative actions related to
	  the adoption of this software by the SquirrelMail Project

2010-02-20  Dave McMurtrie  <davemcmurtrie@gmail.com>
	* common.h: Updated version to 1.2.7

2009-10-16  Dave McMurtrie  <davemcmurtrie@gmail.com>
	* main.c: Applied patch by Jose Luis Tallon to fix compiler warnings.
	  Applied patch by Jose Luis Tallon to provide for default config
	  options.  Applied patch by Jose Luis Tallon to improve server
	  connect retry logic.

	* request.c: Applied patch by Jose Luis Tallon to fix compiler
	  warnings.

	* select.c: Applied patch by Jose Luis Tallon to fix broken syslog
	  call.

	* config.c: Applied patch by Jose Luis Tallon to provide for default
	  config options.  Applied patch by Jose Luis Tallon to improve server
	  connect retry logic. 

	* imapproxy.h: Applied patch by Jose Luis Tallon to provide for default
	  config options.  Applied patch by Jose Luis Tallon to improve server
	  connect retry logic.

2009-01-12  Dave McMurtrie  <davemcmurtrie@gmail.com>
	* main.c: Applied patch by Michael M. Slusarz to add XIMAPPROXY
	  to capability string (fixed patch).

	* common.h: updated version to 1.2.7rc2

2008-10-20  Dave McMurtrie  <davemcmurtrie@gmail.com>
	* request.c: Fixed buffer overflow condition when doing AUTH LOGIN.
	  Applied patch by Michael M. Slusarz to make internal
	  commands RFC compliant (prepend with X instead of P_).  Added
	  support for XPROXYREUSE response.
	  
	* README: Applied patch by Michael M. Slusarz to describe new
	  internal command names (prepended with X instead of P_).


src/config.c
/*
**
** Copyright (c) 2010-2016 The SquirrelMail Project Team
** Copyright (c) 2002-2010 Dave McMurtrie
**
** Licensed under the GNU GPL. For full terms see the file COPYING.
**
** This file is part of SquirrelMail IMAP Proxy.
**
**  Facility:
**
**	config.c
**
**  Abstract:
**
**	Routines for parsing a config file and setting global configuration
**	options.
**
**  Authors:
**
**      Dave McMurtrie <davemcmurtrie@hotmail.com>
**
**  Version:
**
**      $Id$
**
**  Modification History:
**
**      $Log$
**
**      Revision 1.18  2009/10/16 14:34:49  dave64
**      Applied patch by Jose Luis Tallon to improve server connect retry logic.
**
**      Revision 1.17  2009/10/16 14:29:03  dave64
**      Applied patch by Jose Luis Tallon to allow for default config options.
**
**      Revision 1.16  2007/05/31 12:08:32  dave64
**      Applied ipv6 patch by Antonio Querubin.
**
**      Revision 1.15  2005/07/06 12:17:44  dgm
**      Add enable_admin_commands to ConfigTable.
**
**      Revision 1.14  2005/06/15 12:12:26  dgm
**      Patch by Dave Steinberg and Jarno Huuskonen to add chroot_directory
**      config option.
**
**      Revision 1.13  2005/01/12 17:49:51  dgm
**      Applied patch by David Lancaster to provide force_tls config
**      option.
**
**      Revision 1.12  2004/11/10 15:26:25  dgm
**      Explictly NULL terminate all strings that are the result of an strncpy.
**
**      Revision 1.11  2004/10/11 18:01:29  dgm
**      Added foreground mode configuration option.
**
**      Revision 1.10  2004/02/24 14:57:47  dgm
**      Added new config option 'enable_select_cache'.
**
**      Revision 1.9  2003/11/14 15:03:58  dgm
**      Patch by Geoffrey Hort <g.hort@unsw.edu.au> to allow
**      configurable listen address.
**
**      Revision 1.8  2003/10/23 06:18:58  dgm
**      Fixed bug in SetBooleanValue doing upcase of Value.
**
**      Revision 1.7  2003/10/09 12:36:08  dgm
**      Added the ability to set boolean configuration options.
**
**      Revision 1.6  2003/05/20 18:42:04  dgm
**      comment changes only.
**
**      Revision 1.5  2003/05/06 12:10:37  dgm
**      Applied patches by Ken Murchison to include additional config options
**      for SSL support.
**
**      Revision 1.4  2003/04/16 12:13:34  dgm
**      Added commodore logo ascii-art comment at the end.
**      Added support for syslog configuration options.
**
**      Revision 1.3  2003/02/17 14:08:42  dgm
**      added an fclose() that I forgot.
**
**      Revision 1.2  2003/01/27 13:59:15  dgm
**      Patch by Gary Mills <mills@cc.UManitoba.CA> to allow compilation
**      using Sun's cc instead of gcc.
**
**      Revision 1.1  2002/12/17 14:26:49  dgm
**      Initial revision
**
*/


#define _REENTRANT
#define MAX_KEYWORD_LEN 128

#include <errno.h>
#include <string.h>
#include <strings.h>
#include <syslog.h>
#include <stdlib.h>
#include <stdio.h>
#include <ctype.h>
#include "common.h"
#include "imapproxy.h"


/*
 * External globals
 */
extern ProxyConfig_Struct PC_Struct;


/*
 * Internal prototypes
 */
static void SetStringValue( char *, char **, unsigned int );
static void SetNumericValue( char *, int *, unsigned int );
static void SetBooleanValue( char *, unsigned int *, unsigned int );

/*
 * An array of Config_Structs will need to be allocated, one for
 * every possible keyword/value combination (plus a NULL).  Note that
 * I declare an array of 100, meaning that there's a hard limit of 99
 * possible config values right now.  Currently, there are only 9 in use.
 */
struct Config_Struct
{
    char Keyword[MAX_KEYWORD_LEN]; /* The configuration keyword */
    void (*(SetFunction))();       /* ptr to function used to set the value */
    void *StorageAddress;          /* address to store the value */
};

struct Config_Struct ConfigTable[ 100 ];

/*
 * Macro to populate the above ConfigTable
 */
#define ADD_TO_TABLE( KEYWORD, SETFUNCTION, STA, INDEX ) \
        strncpy( ConfigTable[ INDEX ].Keyword, KEYWORD, MAX_KEYWORD_LEN -1 ); \
        ConfigTable[ INDEX ].Keyword[ MAX_KEYWORD_LEN - 1 ] = '\0'; \
        ConfigTable[ INDEX ].SetFunction = SETFUNCTION; \
        ConfigTable[ INDEX ].StorageAddress = STA; \
        INDEX++;





/*++
 * Function:	SetStringValue
 *
 * Purpose:	Common routine to assign string values in the config
 *		options struct.
 *
 * Parameters:  ptr to string from config file
 *              ptr to char ptr for dynamically allocated storage for string.
 *              int -- line of config file where the string was read from.
 *
 * Returns:     nada -- exit()s on errors.
 *
 * Authors:     Dave McMurtrie <davemcmurtrie@hotmail.com>
 *
 * Notes:
 *--
 */
static void SetStringValue( char *String, 
			    char **SavedString, 
			    unsigned int linenum )
{
    char *fn = "SetStringValue()";
    unsigned int Size;
    
    Size = strlen( String ) + 1;
    
    /*
     * Do some reasonable bounds checking before we malloc()
     */
    if ( ( Size < 1 ) || ( Size > 4096 ) )
    {
	syslog( LOG_ERR, "%s: Length of string value at line %d of config file is not within size boundaries -- Exiting.", fn, linenum );
	exit( 1 );
    }
    
    *SavedString = malloc( Size );
    
    if ( ! *SavedString )
    {
	syslog( LOG_ERR, "%s: malloc() failed: %s -- Exiting.", fn,
		strerror( errno ) );
	exit( 1 );
    }
    
    memcpy( *SavedString, String, Size );
    
    return;
}


    


/*++
 * Function:    SetNumericValue
 *
 * Purpose:     Common routine to convert an ascii string to a numeric value
 *              and set a config value accordingly.
 *
 * Parameters:  ptr to the string value to convert. 
 *              ptr to int.  (where to store the converted value)
 *              int -- line of config file where the string was read from.
 *
 * Returns:     nada -- exit()s on failure.
 *
 * Authors:     Dave McMurtrie <davemcmurtrie@hotmail.com>
 *
 * Notes:
 *--
 */
static void SetNumericValue( char *StringValue, 
			     int *Value, 
			     unsigned int linenum )
{
    char *fn = "SetNumericValue()";

    /*
     * Need to generalize this routine.  atoi() seems to only set errno
     * on Solaris, making this very platform specific.
     */
    *Value = atoi( (const char *)StringValue );

    if ( ( ! *Value ) &&
	 ( errno == EINVAL ) )
    {
	syslog( LOG_ERR, "%s: numeric value specified at line %d of config file is invalid -- Exiting.", fn, linenum );
	exit( 1 );
    }

    return;
}




/*++
 * Function:    SetBooleanValue
 *
 * Purpose:     Common routine to assign true/false (or yes/no) values in
 *              the config options struct. 
 *
 * Parameters:  ptr to string value from configfile (yes, no, true, false)
 *              ptr to unsigned int -- where to store the boolean value (1/0)
 *              unsigned int -- Config file line number
 *
 * Returns:     nada -- exit()s on errors.
 *
 * Authors:     Dave McMurtrie <davemcmurtrie@hotmail.com>
 *
 * Notes:       This function is not case sensitive.
 *              It will parse any of y, yes, true and store it as 1.
 *              It will parse any of n, no, false and store it as 0.
 *
 *              Major note -- the Value passed into here will be upcased.
 *--
 */
static void SetBooleanValue( char *Value,
			     unsigned int *StoredValue,
			     unsigned int linenum )
{
    char *fn = "SetBooleanValue()";
    char *CP;
    
    /*
     * Upcase for ease of comparison
     */
    CP = Value;
    while( *CP )
    {
	*CP = toupper( *CP );
	CP++;
    }
    
    if ( ( ( Value[0] == 'Y' ) && ( Value[1] == '\0' ) ) ||
	 ( ( Value[0] == 'Y' ) && ( Value[1] == 'E' ) && ( Value[2] == 'S' ) ) )
    {
	*StoredValue = 1;
	return;
    }
    
    if ( ( ( Value[0] == 'N' ) && ( Value[1] == '\0' ) ) ||
	 ( ( Value[0] == 'N' ) && ( Value[1] == 'O' ) ) )
    {
	*StoredValue = 0;
	return;
    }

    if ( !strcmp( "TRUE", Value ) )
    {
	*StoredValue = 1;
	return;
    }
    
    if ( !strcmp( "FALSE", Value ) )
    {
	*StoredValue = 0;
	return;
    }
    
    syslog( LOG_WARNING, "%s: Invalid boolean value '%s' specified at line %d of config file.  Defaulting to FALSE.", fn, Value, linenum );
    *StoredValue = 0;
    return;
}


/*++
 * Function:	SetDefaultConfigValues
 *
 * Purpose:	Set global configuration default values
 *
 * Parameters:	pointer to global config struct
 *
 * Returns:	nada.
 *
 * Authors:	Jose Luis Tallon <jltallon@adv-solutions.net>
 *--
 */
void SetDefaultConfigValues(ProxyConfig_Struct *PC_Struct)
{
    PC_Struct->server_connect_retries = DEFAULT_SERVER_CONNECT_RETRIES;
    PC_Struct->server_connect_delay = DEFAULT_SERVER_CONNECT_DELAY;
    PC_Struct->ipversion = 0;
    PC_Struct->dnsrr = 0;

    return;
}

/*++
 * Function:	SetConfigOptions
 *
 * Purpose:	Set global configuration options by reading and parsing
 *		the configuration file.
 *
 * Parameters:	char pointer to config filename path.
 *
 * Returns:	nada.  exit()s on any error.
 *
 * Authors:	Dave McMurtrie <davemcmurtrie@hotmail.com>
 *
 * Notes:       Sets values in global ProxyConfig_Struct PC_Struct.
 *--
 */
extern void SetConfigOptions( char *ConfigFile )
{
    FILE *FP;
    char *fn = "SetConfigOptions()";
    char Buffer[1024];
    unsigned int LineNumber;
    unsigned int index;
    unsigned int i;
    char *CP;
    char *Keyword;
    char *Value;
   
    index = LineNumber = 0;

    /*
     * initialize the proxy config struct
     */
    memset( &PC_Struct, 0, sizeof PC_Struct );
    

    /*
     * Build our config option table.
     */
    ADD_TO_TABLE( "server_hostname", SetStringValue, 
		  &PC_Struct.server_hostname, index );

    ADD_TO_TABLE( "listen_port", SetStringValue, 
		  &PC_Struct.listen_port, index );

    ADD_TO_TABLE( "listen_address", SetStringValue,
		  &PC_Struct.listen_addr, index );

    ADD_TO_TABLE( "server_port", SetStringValue, 
		  &PC_Struct.server_port, index );

    ADD_TO_TABLE( "connect_retries", SetNumericValue,
		  &PC_Struct.server_connect_retries, index );
    ADD_TO_TABLE( "connect_delay", SetNumericValue,
		  &PC_Struct.server_connect_delay, index );

    ADD_TO_TABLE( "cache_size", SetNumericValue, 
		  &PC_Struct.cache_size, index );

    ADD_TO_TABLE( "cache_expiration_time", SetNumericValue, 
		  &PC_Struct.cache_expiration_time, index );

    ADD_TO_TABLE( "preauth_command", SetStringValue,
		  &PC_Struct.preauth_command, index );

    ADD_TO_TABLE( "auth_sasl_plain_username", SetStringValue,
		  &PC_Struct.auth_sasl_plain_username, index );
    
    ADD_TO_TABLE( "auth_sasl_plain_password", SetStringValue,
		  &PC_Struct.auth_sasl_plain_password, index );

    ADD_TO_TABLE( "auth_shared_secret", SetStringValue,
		  &PC_Struct.auth_shared_secret, index);

    ADD_TO_TABLE( "proc_username", SetStringValue,
		  &PC_Struct.proc_username, index );
    
    ADD_TO_TABLE( "proc_groupname", SetStringValue,
		  &PC_Struct.proc_groupname, index );
    
    ADD_TO_TABLE( "stat_filename", SetStringValue,
		  &PC_Struct.stat_filename, index );

    ADD_TO_TABLE( "syslog_facility", SetStringValue,
		  &PC_Struct.syslog_facility, index );
    
    ADD_TO_TABLE( "protocol_log_filename", SetStringValue,
		  &PC_Struct.protocol_log_filename, index );

    ADD_TO_TABLE( "syslog_prioritymask", SetStringValue,
		  &PC_Struct.syslog_prioritymask, index );
    
    ADD_TO_TABLE( "tls_ca_file", SetStringValue,
		  &PC_Struct.tls_ca_file, index );
    
    ADD_TO_TABLE( "tls_ca_path", SetStringValue,
		  &PC_Struct.tls_ca_path, index );
    
    ADD_TO_TABLE( "tls_cert_file", SetStringValue,
		  &PC_Struct.tls_cert_file, index );
    
    ADD_TO_TABLE( "tls_key_file", SetStringValue,
		  &PC_Struct.tls_key_file, index );

    ADD_TO_TABLE( "tls_ciphers", SetStringValue,
		  &PC_Struct.tls_ciphers, index );

    ADD_TO_TABLE( "tls_verify_server", SetBooleanValue,
		  &PC_Struct.tls_verify_server, index );

    ADD_TO_TABLE( "tls_no_tlsv1", SetBooleanValue,
		  &PC_Struct.tls_no_tlsv1, index );

    ADD_TO_TABLE( "tls_no_tlsv1.1", SetBooleanValue,
		  &PC_Struct.tls_no_tlsv1_1, index );

    ADD_TO_TABLE( "tls_no_tlsv1.2", SetBooleanValue,
		  &PC_Struct.tls_no_tlsv1_2, index );

    ADD_TO_TABLE( "send_tcp_keepalives", SetBooleanValue,
		  &PC_Struct.send_tcp_keepalives, index );

    ADD_TO_TABLE( "chroot_directory", SetStringValue,
		  &PC_Struct.chroot_directory, index );

    ADD_TO_TABLE( "enable_select_cache", SetBooleanValue,
		  &PC_Struct.enable_select_cache, index );

    ADD_TO_TABLE( "foreground_mode", SetBooleanValue,
		  &PC_Struct.foreground_mode, index );

    ADD_TO_TABLE( "force_tls", SetBooleanValue,
		  &PC_Struct.force_tls, index );

    ADD_TO_TABLE( "enable_admin_commands", SetBooleanValue,
		  &PC_Struct.enable_admin_commands, index );
    
    ADD_TO_TABLE( "ipversion_only", SetNumericValue,
		  &PC_Struct.ipversion, index );
    
    ADD_TO_TABLE( "dns_rr", SetBooleanValue,
		  &PC_Struct.dnsrr, index );
    
    ConfigTable[index].Keyword[0] = '\0';
    
    FP = fopen( ConfigFile, "r" );
    
    if ( !FP )
    {
	syslog(LOG_ERR, "%s: Unable to open config file '%s': %s -- Exiting",
	       fn, ConfigFile, strerror( errno ) );
	exit( 1 );
    }
    
    for ( ;; )
    {
	if ( !fgets( Buffer, sizeof Buffer, FP ) )
	    break;
	    
	LineNumber++;
	    
	/*
	 * Nullify comments, and CRLFs
	 */
	CP = strchr( Buffer, '#' );
	if ( CP ) *CP = 0;
	
	CP = strchr( Buffer, '\n' );
	if ( CP ) *CP = 0;
	
	CP = strchr( Buffer, '\r' );
	if ( CP ) *CP = 0;
	
	/*
	 * Any line that started with a comment or CRLF, we'll
	 * skip.
	 */
	if ( !strlen( Buffer ) )
	    continue;
	
	CP = strtok( Buffer, " " );
	if ( !CP )
	{
	    syslog(LOG_ERR, "%s: parse error reading config file at line %d.  Exiting.", fn, LineNumber );
	    exit( 1 );
	}
	
	Keyword = CP;
	
	CP = strtok( NULL, " " );
	
	if ( !CP )
	{
	    syslog(LOG_ERR, "%s: parse error reading config file at line %d.  Exiting.", fn, LineNumber );
	    exit( 1 );
	}
	
	Value = CP;

	// we don't just want the next token, we want the rest of the line
	// (put back the space that strtok() changed into a null character)
	//
	Value[ strlen( Value ) ] = ' ';

	// however, we then have to be careful to remove trailing whitespace
	//
	i = strlen( Value ) - 1;
	while ( ( Value[ i ] == ' ' )
	     || ( Value[ i ] == '\t' )
	     || ( Value[ i ] == '\r' )
	     || ( Value[ i ] == '\n' ) )
	{
	    i--;
	}
	if ( i < ( strlen( Value ) - 1 ) )
	    Value[ i + 1 ] = '\0';
	
	for (i = 0; ConfigTable[i].Keyword[0] != '\0'; i++ )
	{
	    if ( ! strcasecmp( (const char *)Keyword, ConfigTable[i].Keyword ) )
	    {
		( ConfigTable[i].SetFunction )( Value, 
						ConfigTable[i].StorageAddress,
						LineNumber );
		break;
	    }
	    
	}
	
	/*
	 * If we get here and we're at our NULL value at the end of our
	 * keyword array, we cycled through the entire array and never
	 * matched any keyword.
	 */
	if ( ! ConfigTable[i].Keyword )
	{
	    syslog( LOG_ERR, "%s: unknown keyword '%s' found at line %d of config file -- Exiting.", fn, Keyword, LineNumber );
	    exit( 1 );
	}
	
    }

    fclose( FP );
}


/*
 *                            _________
 *                           /        |
 *                          /         |
 *                         /    ______|
 *                        /    /       ________
 *                       |    |        |      /
 *                       |    |        |_____/
 *                       |    |        ______
 *                       |    |        |     \
 *                       |    |        |______\
 *                        \    \_______
 *                         \           |
 *                          \          |
 *                           \_________|
 */
